core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java

```diff
@@ -866,6 +866,12 @@ private static void matchFilter(SubQueryRemoveRule rule,
           LogicVisitor.find(RelOptUtil.Logic.TRUE, ImmutableList.of(c), e);
       final Set<CorrelationId>  variablesSet =
           RelOptUtil.getVariablesUsed(e.rel);
+      // Only consider the correlated variables which originated from this sub-query level.
+      if (!filter.getVariablesSet().isEmpty()) {
+        // Only retain when filter has non-empty correlated variables set to avoid removing
+        // all the correlated variables from parent scope.
```

Comment:
This comment explains why the `if (...)` exists, how about moving it above the `if (...)`.